rfc/rfc-77/rfc-77.md

```diff
@@ -0,0 +1,247 @@
+chosen will depend on the cardinality of the `record-key` and `secondary-key` columns).
+
+| Column Name | Column Type |
+|-------------|-------------|
+| uuid        | string      |
+| ts          | long        |
+| rider       | string      |
+| city        | string      |
+
+`uuid` is the `record-key` column and a SI is built on the `city` column. Few details to be noted:
+1. Base files in the MDT are sorted (HFile) by the 'record-key' for the records stored in the MDT. For the MDT 
+partition hosting SI, records are sorted by `secondary-key`.
+2. Base file will never contain a tombstone record (record mergers never write tombstone or deleted records 
+into base file)
+
+Base File Entries:
+```
+chennai -> c8abbe79-8d89-47ea-b4ce-4d224bae5bfa
+los-angeles -> 9909a8b1-2d15-4d3d-8ec9-efc48c536a01
+los-angeles -> 9809a8b1-2d15-4d3d-8ec9-efc48c536a01
+sfo -> 334e26e9-8355-45cc-97c6-c31daf0df330
+sfo -> 334e26e9-8355-45cc-97c6-c31daf0df329
+```
+
+Log File Entries:
+```
+sfo -> (334e26e9-8355-45cc-97c6-c31daf0df329, deleted)
+chennai -> e3cf430c-889d-4015-bc98-59bdce1e530c
+los-angeles -> (9809a8b1-2d15-4d3d-8ec9-efc48c536a01, deleted)
+austin -> 9809a8b1-2d15-4d3d-8ec9-efc48c536a01
+```
+
+Merged Records (sorted):
+``` 
+austin -> 9809a8b1-2d15-4d3d-8ec9-efc48c536a01
+chennai -> e3cf430c-889d-4015-bc98-59bdce1e530c
+chennai -> c8abbe79-8d89-47ea-b4ce-4d224bae5bfa
+los-angeles -> 9909a8b1-2d15-4d3d-8ec9-efc48c536a01
+sfo -> 334e26e9-8355-45cc-97c6-c31daf0df330
+```
+
+An important detail to be noted here is that similarity depends on the `secondary-key, record-key` tuple (`city, uuid` 
+in this example). This is radically different from any other existing partitions in Hudi (MDT or base tables). Hence, 
+support needs to be added to merge records from MDT partition hosting SI (with generalized approach that can be used for
+any future non-unique record key based dataset).
+
+The data structure built for holding the merged log records should enable to search similar keys efficiently with 
+following consideration:
+1. Should allow spilling to disk (to prevent OOM). A pure in-memory data structure will not be production ready, but 
+can be used only for POC
+2. Should allow for searching (and retrieving) based on two keys as efficiently as possible. First search will be based 
+on `secondary-key` and second search will be based on `record-key`. Hence, a single level Map used by 
+`HoodieMergedLogRecordScanner` will not work.
+3. Should allow for efficient removal of records (deleted record need to be removed). Hence, any data structure that 
+uses a flat array will not be efficient. 
+4. Should allow for efficient insertion of records (for inserting merged record and for buffering fresh records). 
+
+The initial POC makes use of an in-memory nested maps - with the first level keyed by `secondary-key`  
+and the second level keyed by`record-key`. The final design should allow spilling to disk. Here are some options:
```

Comment:
What are we leaning towards here? Love to understand the overall base log merge algorithm before we can finalize. 

The issue with reusing ExternalSpillableMap is that it's a map<string,hoodierecord> and it's what we plug into logscanner and mergehandle? What are we planning to change this into. 

A multimap would make this map<string, map<string, hoodierecord>>? Can't we just do a Map<Pair<String,String>, HoodieRecord> and reuse the existing implementation of Spillable map?

Reply:
Updated more details with API changes. Basically proposing to make HoodieKey itself generic enough to support secondary keys. Please review this section again.